input/postgres/explain_analyze_test.go

```diff
@@ -0,0 +1,387 @@
+		err := db.QueryRow("SELECT pganalyze.explain_analyze($1, $2, $3, $4)", pair.query, pq.Array(pair.params), pq.Array(pair.paramTypes), pq.Array(pair.analyzeFlags)).Scan(&output)
+		if err != nil {
+			errStr = fmt.Sprintf("%s", err)
+		}
+
+		// Avoid differences in test runs by masking total planning/execution time stats
+		re := regexp.MustCompile(`("(Planning|Execution) Time":) [\d.]+`)
+		output = re.ReplaceAllString(output, "$1 XXXXX")
+
+		if output != pair.expectedOutput {
+			t.Errorf("Incorrect output for query '%s' (direct):\n got: %s\n expected: %s", pair.query, output, pair.expectedOutput)
+		}
+
+		if errStr != pair.expectedError {
+			t.Errorf("Incorrect error for query '%s' (direct):\n got: %s\n expected: %s", pair.query, errStr, pair.expectedError)
+		}
+	}
+}
+
+type queryRunTestpair struct {
+	query          string
+	params         []null.String
+	paramTypes     []string
+	expectedOutput string
+	expectedError  string
+}
+
+var queryRunTests = []queryRunTestpair{
+	{
+		"SELECT 1",
+		[]null.String{},
+		[]string{},
+		`[
+  {
+    "Plan": {
+      "Node Type": "Result",
+      "Parallel Aware": false,
+      "Async Capable": false,
+      "Actual Rows": 1,
+      "Actual Loops": 1
+    },
+    "Planning Time": XXXXX,
+    "Triggers": [
+    ],
+    "Execution Time": XXXXX
+  }
+]`,
+		"",
+	},
+	{
+		"SELECT pg_reload_conf()",
+		[]null.String{},
+		[]string{},
+		"",
+		"pq: permission denied for function pg_reload_conf",
+	},
+	{
+		"UPDATE test SET id = 123",
+		[]null.String{},
+		[]string{},
+		"",
+		"pq: cannot execute UPDATE in a read-only transaction",
+	},
+	{
+		"SELECT 1; UPDATE test SET id = 123",
+		[]null.String{},
+		[]string{},
+		"",
+		"query is not permitted to run (multi-statement or utility command?)",
+	},
+	{
+		"SELECT $1",
+		/* EXECUTE pganalyze_explain_analyze (*/ []null.String{null.StringFrom("1); SELECT ('data'")}, /* ) */
+		[]string{},
+		`[
+  {
+    "Plan": {
+      "Node Type": "Result",
+      "Parallel Aware": false,
+      "Async Capable": false,
+      "Actual Rows": 1,
+      "Actual Loops": 1
+    },
+    "Planning Time": XXXXX,
+    "Triggers": [
+    ],
+    "Execution Time": XXXXX
+  }
+]`,
+		"",
+	},
+	{
+		"SELECT $1",
+		[]null.String{null.StringFrom("dummy")},
+		/* PREPARE pganalyze_explain_analyze (*/ []string{"text) AS SELECT 'data'; PREPARE dummy (text"}, /*) AS [query] */
+		"",
+		`pq: type "text) AS SELECT 'data'; PREPARE dummy (text" does not exist`,
+	},
+}
+
+func TestExplainAnalyzeForQueryRun(t *testing.T) {
+	db := setupTest(t)
+	defer db.Close()
+
+	for _, pair := range queryRunTests {
+		var errStr string
+		output, err := postgres.RunExplainAnalyzeForQueryRun(context.Background(), db, pair.query, pair.params, pair.paramTypes, "", true)
+		if err != nil {
+			errStr = fmt.Sprintf("%s", err)
+		}
+
+		// Avoid differences in test runs by masking total planning/execution time stats
+		re := regexp.MustCompile(`("(Planning|Execution) Time":) [\d.]+`)
```

Comment:
I like that we're testing this, but the `testMode` argument is pretty gross. If I change the regexp to
```suggestion
               re := regexp.MustCompile(`("(?:Planning Time|Execution Time|Startup Cost|Total Cost|Actual Startup Time|Actual Total Time)":) [\d.]+`)
```
I get what looks like fairly stable output (there's buffer hit info, but it's all zeros, and I think given the trivial queries, we can assume it will stay that way). Thoughts?

Reply:
Yeah, you're right - will revise.